packages/ipa-certified-attributes-importer/src/services/openDataExtractor.ts

```diff
@@ -0,0 +1,255 @@
+/* eslint-disable sonarjs/cognitive-complexity */
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/* eslint-disable @typescript-eslint/array-type */
+import axios from "axios";
+import { z } from "zod";
+import { match } from "ts-pattern";
+import { config } from "../config/config.js";
+import { ORIGIN_IPA } from "../index.js";
+
+type Classification = "Agency" | "AOO" | "UO";
+
+export type InstitutionKind = "Agency" | "AOO" | "UO";
+
+export type Institution = {
+  id: string;
+  originId: string;
+  taxCode: string;
+  category: string;
+  description: string;
+  digitalAddress: string;
+  address: string;
+  zipCode: string;
+  origin: string;
+  kind: string;
+  classification: Classification;
+};
+
+export type Category = {
+  code: string;
+  name: string;
+  kind: string;
+  origin: string;
+};
+
+const institutionsFields = [
+  "Codice_IPA",
+  "Denominazione_ente",
+  "Denominazione_aoo",
+  "Descrizione_uo",
+  "Codice_fiscale_ente",
+  "Codice_Categoria",
+  "Mail1",
+  "Indirizzo",
+  "CAP",
+  "Tipologia",
+  "Codice_uni_aoo",
+  "Codice_uni_uo",
+] as const;
+type InstitutionsFields = (typeof institutionsFields)[number];
+
+const categoriesFields = [
+  "Codice_categoria",
+  "Nome_categoria",
+  "Tipologia_categoria",
+] as const;
+type CategoriesFields = (typeof categoriesFields)[number];
+
+function fieldExtractor<K>(
+  record: unknown[],
+  fields: Map<K, number>
+): <T>(key: K, decoder: z.ZodSchema<T>) => T | undefined {
+  return (key, decoder) => {
+    const valueIndex = fields.get(key);
+    if (!valueIndex) {
+      return undefined;
+    }
+
+    const value = decoder.safeParse(record[valueIndex]);
+    if (!value.success) {
+      return undefined;
+    }
+
+    return value.data;
+  };
+}
+
+export async function getAllCategories(): Promise<Category[]> {
+  const response = await axios.get(config.institutionsCategoriesUrl);
+
+  const fields: Map<CategoriesFields, number> = new Map(
+    (response.data.fields as Array<{ id: string }>)
+      .map<[string, number]>((f, index) => [f.id, index])
+      .filter(([f, _]) => (categoriesFields as readonly string[]).includes(f))
+      .map(([f, i]) => [f as CategoriesFields, i])
+  );
+
+  return (response.data.records as Array<any>).reduce<Category[]>(
+    (accumulator, record: any[]) => {
+      const extractor = fieldExtractor<CategoriesFields>(record, fields);
```

Comment:
Could we use unknown instead of any here?

Reply:
using unknown I have some problems with the reducer that I couldn't resolve. If it's not a huge problem I would like to leave it as it is